### PR TITLE
✨ Add transition_id method to BaseRecord

### PR DESCRIPTION
**Pull request by julien-du**
As part of this ticket : https://app.shortcut.com/wecasa/story/43835/zoho-blueprint-transition-get-transition-dynamically-from-zoho

### CODE REVIEW

File: spec/zoho_hub/base_record_spec.rb [hunk 1]
```diff
@@ -55,6 +41,48 @@
...
+        .to_return(status: 200,
+                   body: { blueprint: { transitions: [{ next_field_value: 'Closed',
+                                                        id: 'transition-123' }] } }.to_json)
+      put_stub = \
+        stub_request(:put, "https://crmsandbox.zoho.eu/crm/v2/Leads/#{test_instance.id}/actions/blueprint")
+        .with(body: { blueprint: [{ transition_id: 'transition-123', data: {} }] })
+        .to_return(status: 200,
+                   body: {}.to_json)
+
+      test_instance.blueprint_transition('Closed', {})
```
Comment (romain-verbeke) [inline spec/zoho_hub/base_record_spec.rb]:
```suggestion
      test_instance.blueprint_transition('Closed')
```
should work too?

**Review by romain-verbeke — COMMENTED** [review]
Very nice idea i think, but you should also make a PR in the web repo that will reference the zoho_hub gem on this branch and fix everything that should break ?

**Comment by julien-du** [comment]
> Very nice idea i think, but you should also make a PR in the web repo that will reference the zoho_hub gem on this branch and fix everything that should break ?

Yes exactly, it's ongoing :)

File: lib/zoho_hub/base_record.rb [hunk 2]
```diff
@@ -228,9 +228,19 @@ def delete_record
       build_response(body)
     end
 
-    def blueprint_transition(transition_id, data = {})
+    def transition_id(next_field_value)
+      response = blueprint_transitions
+      transition = response.data[:blueprint][:transitions].find do |t|
+        t[:next_field_value] == next_field_value
+      end
+      raise "Unknown Account Blueprint transition: #{next_field_value}" unless transition
```
Comment (jlestavel) [inline lib/zoho_hub/base_record.rb]: Can you please define an error class for this please?

File: spec/zoho_hub/base_record_spec.rb [hunk 3]
```diff
@@ -55,6 +41,48 @@
...
+        expect(record.my_string).to eq('')
+        expect(record.my_bool).to eq(false)
+      end
+    end
+  end
+
+  describe '#blueprint_transition' do
+    let(:test_instance) do
+      described_class.new
+    end
```
Comment (jlestavel) [inline spec/zoho_hub/base_record_spec.rb]:
```suggestion
    let(:test_instance) { described_class.new }
```

File: spec/zoho_hub/base_record_spec.rb [hunk 4]
```diff
@@ -55,6 +41,48 @@
     end
   end
 
+  describe '#build_response' do
+    context 'with an empty string and a "false" boolean' do
+      let(:body) { { data: [{ My_String: '', My_Bool: false }] } }
+
+      it 'correctly construct the record' do
```
Comment (jlestavel) [inline spec/zoho_hub/base_record_spec.rb]:
```suggestion
      it 'correctly builds the record' do
```

File: spec/zoho_hub/base_record_spec.rb [hunk 5]
```diff
@@ -55,6 +41,48 @@
...
+      get_stub = \
+        stub_request(:get, "https://crmsandbox.zoho.eu/crm/v2/Leads/#{test_instance.id}/actions/blueprint")
+        .to_return(status: 200,
+                   body: { blueprint: { transitions: [{ next_field_value: 'Closed',
+                                                        id: 'transition-123' }] } }.to_json)
+      put_stub = \
+        stub_request(:put, "https://crmsandbox.zoho.eu/crm/v2/Leads/#{test_instance.id}/actions/blueprint")
+        .with(body: { blueprint: [{ transition_id: 'transition-123', data: {} }] })
+        .to_return(status: 200,
+                   body: {}.to_json)
```
Comment (jlestavel) [inline spec/zoho_hub/base_record_spec.rb]: Couldn't we define those as `let!`?

File: spec/zoho_hub/base_record_spec.rb [hunk 6]
```diff
@@ -55,6 +41,48 @@
...
+  end
+
+  describe '#blueprint_transition' do
+    let(:test_instance) do
+      described_class.new
+    end
+
+    before do
+      allow(described_class).to receive(:to_s).and_return('Leads')
+      allow(test_instance).to receive(:id).and_return('123456789')
```
Comment (jlestavel) [inline spec/zoho_hub/base_record_spec.rb]: Can't you set that during the `test_instance` definition? (`described_class.new(id: '123456789')`)

File: spec/zoho_hub/base_record_spec.rb [hunk 7]
```diff
@@ -55,6 +41,48 @@
...
+    end
+  end
+
+  describe '#blueprint_transition' do
+    let(:test_instance) do
+      described_class.new
+    end
+
+    before do
+      allow(described_class).to receive(:to_s).and_return('Leads')
```
Comment (jlestavel) [inline spec/zoho_hub/base_record_spec.rb]: Shouldn't you define a `test_class` instead? (like they do line 4)

**Review by jlestavel — CHANGES_REQUESTED** [review]

**Review by jlestavel — APPROVED** [review]

**Pull request merged**